packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/mongo/MongoCollection.kt

```diff
@@ -0,0 +1,236 @@
+    return (this as MongoCollection<T?, BsonValue>).findOne(filter, projection, sort)
+}
+
+public suspend inline fun <reified T, K : Any> MongoCollection<T, K>.find(filter: BsonDocument? = null, projection: BsonDocument? = null, sort: BsonDocument? = null, limit: Long? = null): List<T> {
+    isType<MongoCollectionImpl<*, *>>(this)
+    return decodeFromBsonValueList(find(filter, projection, sort, limit).asArray().toList())
+}
+
+@JvmName("findTyped")
+public suspend inline fun <reified T> MongoCollection<*, *>.find(filter: BsonDocument? = null, projection: BsonDocument? = null, sort: BsonDocument? = null, limit: Long? = null): List<T> {
+    return (this as MongoCollection<T, BsonValue>).find(filter, projection, sort, limit)
+}
+
+public suspend inline fun <reified T, K : Any> MongoCollection<T, K>.aggregate(pipeline: List<BsonDocument>): List<T> {
+    isType<MongoCollectionImpl<*, *>>(this)
+    return decodeFromBsonValueList(aggregate(pipeline))
+}
+
+@JvmName("aggregateTyped")
+public suspend inline fun <reified T> MongoCollection<*, *>.aggregate(pipeline: List<BsonDocument>): List<T> {
+    return (this as MongoCollection<T, BsonValue>).aggregate(pipeline)
+}
+
+public suspend inline fun <reified T : Any, reified R : Any> MongoCollection<T, R>.insertOne(document: T): R {
+    isType<MongoCollectionImpl<*, *>>(this)
+    return decodeFromBsonValue(insertOne(encodeToBsonValue(document).asDocument()))
+}
+
+@JvmName("insertOneTyped")
+public suspend inline fun <reified T : Any, reified R : Any> MongoCollection<*, *>.insertOne(document: T): R {
+    return (this as MongoCollection<T, R>).insertOne(document)
+}
+
+public suspend inline fun <reified T : Any, reified R : Any> MongoCollection<T, R>.insertMany(
+    documents: Collection<T>,
+): List<R> {
+    isType<MongoCollectionImpl<*, *>>(this)
+    return decodeFromBsonValueList(insertMany(documents.map { encodeToBsonValue(it).asDocument() }))
+}
+
+@JvmName("insertManyTyped")
+public suspend inline fun <reified T : Any, reified R : Any> MongoCollection<*, *>.insertMany(documents: Collection<T>): List<R> {
+    return (this as MongoCollection<T, R>).insertMany(documents)
+}
+
+public suspend fun MongoCollection<*, *>.deleteOne(filter: BsonDocument): Boolean {
+    isType<MongoCollectionImpl<*, *>>(this)
+    return deleteOne(filter)
+}
+
+public suspend fun MongoCollection<*, *>.deleteMany(filter: BsonDocument): Long {
+    isType<MongoCollectionImpl<*, *>>(this)
+    return deleteMany(filter)
+}
+
+// FIXME Could just return Boolean, since matchedCount=1,modifiedCount=1 even if multiple documents should be matching :thinking:
+// FIXME Should we split into upsertOne, since response only contains 'upsertedId' if call has 'upsert:true`
+public suspend inline fun <T : Any, reified R> MongoCollection<T, R>.updateOne(
+    filter: BsonDocument,
+    update: BsonDocument,
+    upsert: Boolean = false
+): Pair<Boolean, R?> {
```

Comment:
I'm not really a fan of using Pair in the public API since accessors are not very informative (first, second)... Should we try to use the same `UpdateResult` class as the driver do: https://github.com/mongodb/mongo-java-driver/blob/master/driver-core/src/main/com/mongodb/client/result/UpdateResult.java#L31

I do realize that stuff like `matchedCount` and `modifiedCount` is a bit annoying for the updateOne case, but Pair also feels too generic....I guess we could introduce an `UpdateOneResult` and `UpdateResult`? But maybe it is better to just follow the driver on this one 🤔

Reply:
I went with a data classes for each of the operations.